OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs

```diff
@@ -32,41 +32,53 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		[Desc("Are upgrades irrevocable once the conditions have been met?")]
 		public readonly bool GrantPermanently = false;
 
-		public object Create(ActorInitializer init) { return new UpgradeOnDamage(init.Self, this); }
+		public object Create(ActorInitializer init) { return new UpgradeOnDamageState(init.Self, this); }
 	}
 
-	public class UpgradeOnDamage : INotifyDamageStateChanged
+	public class UpgradeOnDamageState : INotifyDamageStateChanged, INotifyCreated
 	{
-		readonly UpgradeOnDamageInfo info;
+		readonly UpgradeOnDamageStateInfo info;
 		readonly UpgradeManager um;
+		readonly Health health;
 		bool granted;
 
-		public UpgradeOnDamage(Actor self, UpgradeOnDamageInfo info)
+		public UpgradeOnDamageState(Actor self, UpgradeOnDamageStateInfo info)
```

Comment:
Do we want an upgrade rule for this? This trait has never been released, but it seems that there are already people using it.